### ISSUE TITLE
Inconsistent data with prometheus between refreshes

### ISSUE THREAD

**Issue by rdubya16** [coreos/prometheus-operator#2203]
**What did you do?**
Installed kube-prometheus on cluster with ingress

**What did you expect to see?**
Consistent data with each refresh of prometheus alert output

**What did you see instead? Under which circumstances?**
Alert/Graph data alternating between refreshes

**Environment**
EKS

* Prometheus Operator version:
v0.25.0

* Kubernetes version information:
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-28T15:20:58Z", GoVersion:"go1.11", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.11-eks", GitCommit:"6bf27214b7e3e1e47dce27dcbd73ee1b27adadd0", GitTreeState:"clean", BuildDate:"2018-12-04T13:33:10Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
* Kubernetes cluster kind: 
eks

I am seeing inconsistent output from prometheus alarm and graph data. Each refresh will toggle back and forth as it round robins between each instance. An alarm will show on one instance but not the other. Is this normal behaviour? I found this issue here where it seems to indicate I might be using the wrong service? https://github.com/coreos/prometheus-operator/pull/113

The ingress im using is identical to the one generated by the jsonnet build when enabling ingress as per the documentation. Thanks!

**Comment by metalmatze** [comment]
That sounds like you're running more than 1 replicas of Prometheus. Our default in the stack is 2. I'm guessing that your ingress is doing a round robin between those replicas. You might be able to set something like a sticky session with your ingress.

**Comment by rdubya16** [comment]
@metalmatze I am running two replicas as per the kube-prometheus defaults. What are general best practices in this case? Just run one instance and let k8s recover it in the event of a failure?

**Comment by metalmatze** [comment]
It depends on your use case, I'd say. 

Running 1 replica is one possibility of course. But as I wrote you can also have sticky session with your Ingress so it always routes to one of the replicas. 
Another option could be to have 2 paths and then route to each Prometheus. I know some people do this for Federation with Prometheus. The later needs some adjustments with the Service and Ingress.

**Comment by brancz** [comment]
Yes sticky session is what I recommend. The Thanos project tries to address this particular issue in more detail, allowing you to either have Thanos apply heuristics to "deduplicate" or have it show you the result from both Prometheus servers simultaneously.

**Comment by NarendranathPanda** [comment]
Hi , 
 We are facing the same issue, We are using Prometheus Operator with replicas 2,  We are able to fix it by changing the replicas to 1 . Is there any other better solution

**Comment by brancz** [comment]
@NarendranathPanda in the Service in front of Prometheus you can configure `sessionAffinity: ClientIP`, then the same user will hit the same Prometheus server on consecutive requests.

**Comment by laurivosandi** [comment]
Prometheus Operator creates the `prometheus-operated` service, however I can't seem to find an easy way to set the session affinity to client there. What about making `ClientIP` the default there?
